kipoi/postprocessing/variant_effects/mutation_map.py

```diff
@@ -0,0 +1,683 @@
+                            mutmap_dict = {}
+                            preds = predictions[scoring_fn][model_output].loc[predictions_rowfilt]
+                            # append predictions for ref/ref variants:
+                            ref_ref_preds = pd.Series(0, name= preds.name, index=ref_ref_var_ids)
+                            preds = pd.concat([preds, ref_ref_preds])
+                            preds = preds.loc[correct_order_ids].values
+                            mutmap_dict["mutation_map"] = np.reshape(preds, (metadata_seqlen, 4)).T
+                            mutmap_dict["ovlp_var"] = query_vcf_recs_rel
+                            mutmap_dict["ref_seq"] = ref_seq_here
+                            mutmap_dict["metadata_region"] = metadata_subset_dict
+                            metadata_mutmap[scoring_fn][model_output] = mutmap_dict
+                    for seq_key in self.meta_to_seq[metadata_key]:
+                        pl_mut_maps[seq_key] = metadata_mutmap
+                mutation_map.append(pl_mut_maps)
+        self.mutation_map = mutation_map
+        return mutation_map
+
+    def save_to_file(self, fname):
+        # Best is to use a hdf5 file
+        res = self.get_merged_data()
+        res = {"_list_%d"%i:v for i, v in enumerate(res)}
+        ofh = h5py.File(fname, "w")
+        recursive_h5_mutmap_writer(res, ofh)
+        ofh.close()
+
+
+class MutationMapDrawer(object):
+    def __init__(self, mutation_map=None, fname=None):
+        if mutation_map is None and fname is None:
+            raise Exception("Either mutation_map or fname for a mutation_map file has to be given")
+        if mutation_map is not None:
+            self.mutation_map = mutation_map
+        elif fname is not None:
+            with h5py.File(fname, "r") as ifh:
+                mutation_map = recursive_h5_mutmap_reader(ifh)
+                mutation_map = [mutation_map["_list_%d" % i] for i in range(len(mutation_map))]
+                self.mutation_map = mutation_map
+
+
+    def draw_mutmap(self, dl_entry, model_seq_key, scoring_key, model_output, ax=None, show_letter_scale = False,
+                    cmap = plt.cm.bwr, limit_region = None):
+        from .utils.seqplotting_deps import encodeDNA
+        mm_obj = self.mutation_map[dl_entry][model_seq_key][scoring_key][model_output]
+
+        # Derive letter heights from the mutation scores.
+        letter_heights = encodeDNA([mm_obj["ref_seq"]])[0, ...]
+        # letter_heights = letter_heights * np.abs(mm_obj['mutation_map'].sum(axis=0))[:,None]
+        letter_heights = letter_heights * np.abs(mm_obj['mutation_map'].mean(axis=0))[:, None]
+
+        return seqlogo_heatmap(letter_heights, mm_obj['mutation_map'], mm_obj['ovlp_var'], vocab="DNA", ax = ax,
+                               show_letter_scale = show_letter_scale, cmap=cmap, limit_region =limit_region)
+
+
+
+
+def recursive_h5_mutmap_writer(objs, handle, path =""):
```

Comment:
could you re-use https://github.com/kipoi/kipoi/blob/master/kipoi/writers.py#L123 to prevent code duplication? The writer is pretty general.

Reply:
don't see how, my `recursive_h5_mutmap_writer` takes lists, dicts and single values / numpy arrays as inputs. if lists or dicts are given they are recursed up to the the leaf elements. that is more general than `kipoi.writers.HDF5BatchWriter` is and it is necessary here. I can transform the `recursive_h5_mutmap_writer` and `recursive_h5_mutmap_reader` methods in a IO class though if you prefer - it has nothing to do really with the `kipoi.writers.HDF5BatchWriter` so I don't think it would make sense to integrate it there.

Reply:
`kipoi.writers.HDF5BatchWriter`/`kipoi.readers.HDF5Reader` also works with any nested combination of numpy arrays (list, dict, list of dicts etc). I solve this problem by 'flattening' the nested structure to a single dictionary [line](https://github.com/kipoi/kipoi/blob/master/kipoi/writers.py#L152) and then un-flattening the dict to a nested structure in the reader - see [these lines](https://github.com/kipoi/kipoi/blob/master/kipoi/readers.py#L54-L71)